rotatable_file.go

```diff
@@ -0,0 +1,95 @@
+package goolog2
+
+import (
+	"os"
+	"strconv"
+	"time"
+)
+
+type rotableFile struct {
+	simpleFile
+	filePath      string
+	maxSize       int64
+	checkInterval time.Duration
+	nextCheck     time.Time
+	timesrc       TimeSource
+}
+
+func NewRotatableFile(
+	timesrc TimeSource,
+	filepath string,
+	sync bool,
+	maxSize int64,
+	checkInterval time.Duration,
+) FileHolder {
+	holder := &rotableFile{}
+	holder.simpleFile.init(filepath, sync)
+	if maxSize > 0 {
+		holder.maxSize = maxSize
+		holder.filePath = filepath
+		holder.checkInterval = checkInterval
+		holder.timesrc = timesrc
+		holder.nextCheck = timesrc.Now().Add(checkInterval)
+	}
+	return holder
+}
+
+func (this *rotableFile) Ref() FileHolder {
+	this.simpleFile.Ref()
+	return this
+}
+
+func (this *rotableFile) AccessWriter(
+	functor func(writer FileWriter),
+) {
+	if this.writer != nil {
+		/* --  The lock avoids inter-mixing of logging lines */
+		this.mutex.Lock()
+		defer this.mutex.Unlock()
+		functor(this.writer)
+		if this.sync {
+			this.writer.Sync()
+		}
+		if this.needRotate() {
+			this.rotate()
+		}
```

Comment:
V patternFile mám rotaci pod vlastním zámkem. To bylo záměrně, protože jsem nechtěl ostatní vlákna zdržovat rotováním, které může být docela náročné. V tvém případě se dělá hned celé skupina operací nad filesystémem - voláš sekvenci statů a pak následně řadu přejmenování souborů.

Zvážil bych úpravu, zda rotování nedělat v samostatném vláknu. A bylo by hezké tuhle úpravu přenést i do patternFile.